### PR TITLE
CMEK shoot cluster manual test documentation

### LINKED ISSUE #78: Enhancement to support CMEK Encrypted Disks
**What would you like to be added**:
1. End to end changes required to support CMEK Encrypted disks. See source requirement at: https://github.com/gardener/gardener-extension-provider-gcp/issues/564 

**Why is this needed**:
To support  https://github.com/gardener/gardener-extension-provider-gcp/issues/564  for ITAR

####  Machine Class Changes
- [x] Addition of `kmsKeyName` and in the provider spec to map inside `kmsKeyName` in GCP `CustomerEncryptionKey`
   Machine Controller Manager GCP Provider Changes. There is no need for separate specification of `kmsKeyServiceAccountName` since the current `GCPProviderSpec` already has a `ServiceAccounts []GCPServiceAccount` and introducing another would be redundant. (NOTE: Clarify why is this a slice since technically a VM can be associated with only ONE service account)
- [x] Enhancement of `GCPProviderSpec.GCPDisk` with kms key name field. 
- [x] Enhancement of `GCPDisk` such that [AttachedDisks](https://pkg.go.dev/google.golang.org/api/compute/v1#AttachedDisk) are populated using the right encryption key using key name from the machine class, whose origin source is in the shoot yaml.
- [x] Enhancement of `CreateMachine->CreateMachineUtil` to ensure that `compute.Instances` are launched with encrypted disks populating the `CustomerEncryptionKey` struct appropriately using the `kmsKeyName`  and optionally `kmsKeyServiceAccountName` taken from the `GCPProviderSpec`.
- [x] Unit tests
- [x] End to End tests

#### GCP Provider Extension
- [x] Enhance `github.com/gardener/gardener-extension-provider-gcp/pkg/apis/gcp.WorkerConfig` in backward incompatible manner to add fields for both `bootVolume` and `dataVolume`, each containing encryption key names: `kmsKeyName`
- [x] Enhance the machine deployment generation code in `generateMachineConfig`  to generate machine classes with `kmsKeyName` and optionally `kmsKeyServiceAccountName` in the provider spec.
- [x] Unit Tests
- [x] End to End Tests
- [x] Resolve addition of IAM Policy bindings to keys wrt service accounts
- [x] Resolve on whether whether `roles/cloudkms.admin` or `roles/cloudkms.cryptoKeyEncrypterDecrypter` is sufficient. FINDING:  `roles/cloudkms.cryptoKeyEncrypterDecrypter` is sufficient for encrypted disks. `roles/cloudkms.admin` is only needed for creation/deletion and general admin of keys.

#### Timeline

- [x] Gardener extension release (22nd June)
- [x] MCM-provider GCP (16 June) (Done on 8th June)

Live update (28th June) , Canary Freeze (22nd June)

Target the releases from our side -> 16th June

##### Multiple GCP Service Accounts
- [x] Test VM launch with multiple GCP service accounts.
- [x] Test diff KMS keys for diff disks giving privilege to these GCP service accounts. 
- [x] Test with old shoot YAMLs and new shoot YAMLs. Backward compat testing

### PR DESCRIPTION
**Pull request by elankath**
**What this PR does / why we need it**:
Dev Manual Testing Doc for CMEK Updated

**Which issue(s) this PR fixes**:
Fixes:- part of #78

**Special notes for your reviewer**:

**Release note**:

```improvement developer
Enhanced Dev Testing Doc for CMEK
```

### CODE REVIEW

File: docs/development/testing/cmek.md [hunk 1]
```diff
@@ -52,18 +61,112 @@
...
+         kmsKeyServiceAccount: user@projectId.iam.gserviceaccount.com 
+      diskSizeGb: '25'
+      index: 2
+      interface: SCSI
+      kind: compute#attachedDisk
+      mode: READ_WRITE
+      source: <...>
+      type: PERSISTENT
+    ```
+1. One can change the shoot YAML by adding removing volumes and once the shoot has re-conciled and the new machine in `Running` phase, the persistent disks associated wth the machine VM should be correspondingly added/removed and still be configured with the specified encryoption key.
```
Comment (rishabh-11) [inline docs/development/testing/cmek.md:172]: One can change the shoot YAML by adding/removing volumes. Once the shoot has reconciled with the new machines in the `Running` phase, the persistent disks associated with the VM should be added/removed and still be configured with the specified encryption key.

File: docs/development/testing/cmek.md [hunk 2]
```diff
@@ -52,18 +61,112 @@
...
+            providerConfig:
+               apiVersion: gcp.provider.extensions.gardener.cloud/v1alpha1
+               kind: WorkerConfig
+               volume:
+                  interface: "SCSI" # applies only for scratch disks. misleading field name
+                  encryption:
+                  kmsKeyName: "projects/projectId/locations/<zoneName>/keyRings/<keyRingName>/cryptoKeys/alpha"
+                  kmsKeyServiceAccount: "user@projectId.iam.gserviceaccount.com" 
+      ```
+1. Apply the shoot YAML agaisnt the `nodeless` cluster and wait till once the shoot has reconciled and the machine(s) in pool are in `Running` phase.
```
Comment (rishabh-11) [inline docs/development/testing/cmek.md:123]: Apply the shoot YAML against the `nodeless` cluster and wait until the shoot has reconciled and the machine(s) in the worker pool are in `Running` phase.

File: docs/development/testing/cmek.md [hunk 3]
```diff
@@ -52,18 +61,112 @@
...
+1. Use `gardenctl` to target the shooted seed and get the provider machine name and zone
+   ```
+   kubectl get mc -A -o=jsonpath='{.items[*].spec.providerID}'
+   gce:///projectId/zone/shoot--userid--shootName-worker-1-z1-65c6b-9ptm7
+   ```
+1. You can check that the encrypted disks have been created:
+    1. Describe the instance: 
+    `gcloud compute instances describe shoot--userid--shootName-worker-1-z1-65c6b-9ptm7 --zone zone` 
+    1. The persistent disks should look like below with  have a disk encryoption key
```
Comment (rishabh-11) [inline docs/development/testing/cmek.md:132]: The persistent disks should look like the one below with a disk encryption key

**Review by rishabh-11 — CHANGES_REQUESTED** [review]
Looks good. Just some minor changes needed.

File: docs/development/testing/cmek.md (hunk 1 above)
Comment (elankath) [reply docs/development/testing/cmek.md]: done

File: docs/development/testing/cmek.md (hunk 2 above)
Comment (elankath) [reply docs/development/testing/cmek.md]: done

File: docs/development/testing/cmek.md (hunk 3 above)
Comment (elankath) [reply docs/development/testing/cmek.md]: done

**Review by rishabh-11 — APPROVED** [review]
/lgtm

**Pull request merged**